tests/python/pants_test/engine/test_fs.py

```diff
@@ -371,6 +372,79 @@ def test_materialize_directories(self):
         content = f.read()
         self.assertEqual(content, "European Burmese")
 
+  def test_strip_prefix(self):
+    # Set up files:
+
+    relevant_files = (
+      'characters/dark_tower/roland',
+      'characters/dark_tower/susannah',
+    )
+    all_files = (
+      'books/dark_tower/gunslinger',
+      'characters/altered_carbon/kovacs',
+    ) + relevant_files + (
+      'index',
+    )
+
+    with temporary_dir() as temp_dir:
+      safe_file_dump(os.path.join(temp_dir, 'index'), 'books\ncharacters\n')
+      safe_file_dump(
+        os.path.join(temp_dir, "characters", "altered_carbon", "kovacs"),
+        "Envoy",
+        makedirs=True,
+      )
+
+      tower_dir = os.path.join(temp_dir, "characters", "dark_tower")
+      safe_file_dump(os.path.join(tower_dir, "roland"), "European Burmese", makedirs=True)
+      safe_file_dump(os.path.join(tower_dir, "susannah"), "Not sure actually", makedirs=True)
+
+      safe_file_dump(
+        os.path.join(temp_dir, "books", "dark_tower", "gunslinger"),
+        "1982",
+        makedirs=True,
+      )
+
+      snapshot, snapshot_with_extra_files = self.scheduler.capture_snapshots((
+        PathGlobsAndRoot(PathGlobs(("characters/dark_tower/*",)), text_type(temp_dir)),
+        PathGlobsAndRoot(PathGlobs(("**",)), text_type(temp_dir)),
+      ))
+      # Check that we got the full snapshots that we expect
+      self.assertEquals(snapshot.files, relevant_files)
+      self.assertEquals(snapshot_with_extra_files.files, all_files)
+
+      # Strip empty prefix:
+      zero_prefix_stripped_digest = assert_single_element(self.scheduler.product_request(
+        Digest,
```

Comment:
Oh, so I missed the first time around: I think that in the case of both `MergedDirectories` and `PrefixStrippedDirectory`, usages parse pretty strangely:

```
digest = yield Get(Digest, PrefixStrippedDirectory(..))
```

`PrefixStrippedDirectory` "sounds" like the _output_ of stripping the prefix (and `MergedDirectories` sounds like the output of some operation). The upside of the current naming is that you don't need a name for the output... so if we can rename these to sound like inputs, that would work. If we still want to use these names, it feels like we'd want to add _separate_ types to act as inputs.

Reply:
https://github.com/pantsbuild/pants/pull/6635#discussion_r225619148 ;)

Renamed to `DirectoryWithPrefixToStrip` - will do `MergedDirectories` in a separate PR if you're happy with it :)

Reply:
I was going to suggest using `Request`, which it sounds like that's out (which I think I agree with).

`DirectoryWithPrefixToStrip` is a great alternative name. It's quite clear that the `prefix` _will_ be stripped, rather than the datatype already being stripped.

`MergedDirectories` would be great to call something like `DirectoriesToMerge`.

Reply:
> [#6635 (comment)](https://github.com/pantsbuild/pants/pull/6635#discussion_r225619148) ;)
> 
> Renamed to `DirectoryWithPrefixToStrip` - will do `MergedDirectories` in a separate PR if you're happy with it :)

Yea, I think that maybe I thought that we were naming the outputs, rather than the inputs? Sorry for the miscommunication.